crates/solidity/outputs/cargo/tests/src/doc_examples/using_queries.rs

```diff
@@ -0,0 +1,176 @@
+use std::path::Path;
+
+use anyhow::Result;
+use infra_utils::paths::PathExtensions;
+use semver::Version;
+use slang_solidity::kinds::RuleKind;
+use slang_solidity::language::Language;
+use slang_solidity::parse_output::ParseOutput;
+use slang_solidity::query::Query;
+
+fn parse_doc_input_file<T: AsRef<Path>>(path: T) -> Result<ParseOutput> {
+    let input_path =
+        Path::repo_path(Path::new("documentation/public/user-guide/inputs").join(path.as_ref()));
+
+    let input_path = input_path.unwrap_str();
+
+    let source = std::fs::read_to_string(input_path)?;
+    let source = source.trim();
+
+    let language = Language::new(Version::new(0, 8, 0))?;
+
+    Ok(language.parse(RuleKind::SourceUnit, source))
+}
+
+#[test]
+fn using_queries() -> Result<()> {
+    #[allow(unused_variables, clippy::items_after_statements)]
+    {
+        let parse_output = parse_doc_input_file("using-the-cursor.sol")?;
+        // --8<-- [start:creating-a-query]
+        use slang_solidity::query::Query;
+
+        // Any `Cursor` can be used to create a query.
+        let cursor = parse_output.create_tree_cursor();
+
+        let query = Query::parse("[ContractDefinition]").unwrap();
+        let result /* : QueryResultIterator */ = cursor.query(vec![query]);
+        // --8<-- [end:creating-a-query]
+    }
+
+    {
+        let parse_output = parse_doc_input_file("using-the-cursor.sol")?;
+        let cursor = parse_output.create_tree_cursor();
+        // --8<-- [start:listing-contract-names]
+        let mut names = vec![];
+
+        let query = Query::parse("[ContractDefinition ... @name [Identifier] ...]").unwrap();
```

Comment:
I suggest using `?` instead of the `.unwrap()` everywhere, as it adds visual noise unrelated to the concept we are explaining.

Reply:
The snippets should be standalone and adding a `?` would implicitly assume we're in a `<Result as Try>` context; also the `E` in `Result<T,E>` here is `String`, which does not implement standard `Error` trait, so it's not easily convertible in a generic setting (e.g. `anyhow::Result`), thus could mislead the users when they write the final code.

We need to rethink whether we want the string as the error or rather use the nom error, instead. cc @AntonyBlakey 

Since the only error can be due to parsing, this is similar to `Regex::new("...").unwrap()`, where the only error that can happen is at compile-time in this case, so I think it's also justified here.

Reply:
The error will end up as a structured type like in the graph builder:

```
// ----------------------------------------------------------------------------
// Parse errors

/// An error that can occur while parsing a graph DSL file
#[derive(Debug, Error)]
pub enum ParseError {
    #[error("Expected quantifier at {0}")]
    ExpectedQuantifier(Location),
    #[error("Expected '{0}' at {1}")]
    ExpectedToken(&'static str, Location),
    #[error("Expected variable name at {0}")]
    ExpectedVariable(Location),
    #[error("Expected unscoped variable at {0}")]
    ExpectedUnscopedVariable(Location),
    #[error("Invalid regular expression /{0}/ at {1}")]
    InvalidRegex(String, Location),
    #[error("Expected integer constant in regex capture at {0}")]
    InvalidRegexCapture(Location),
    #[error("Invalid query pattern: {}", _0.message)]
    QueryError(#[from] QueryError),
    #[error("Unexpected character '{0}' in {1} at {2}")]
    UnexpectedCharacter(char, &'static str, Location),
    #[error("Unexpected end of file at {0}")]
    UnexpectedEOF(Location),
    #[error("Unexpected keyword '{0}' at {1}")]
    UnexpectedKeyword(String, Location),
    #[error("Unexpected literal '#{0}' at {1}")]
    UnexpectedLiteral(String, Location),
    #[error("Query contains multiple patterns at {0}")]
    UnexpectedQueryPatterns(Location),
    #[error(transparent)]
    Check(#[from] crate::checker::CheckError),
}
```

Reply:
So to clarify, the `String` in `fn parse(text: &str) -> Result<Query, String>` is temporary for now?

In any case, I vote leaving `unwrap()` for the context-insensitivity and lack of ability to propagate without explicit error conversion leading the `?` in the current situation.

Reply:
> we lack of ability to propagate without explicit error conversion

I think the `anyhow::Result` we use everywhere is popular enough/understandable over adding the many `.unwrap()` calls, since we don't explicitly talk about/mention error handling.

Reply:
> So to clarify, the String in fn parse(text: &str) -> Result<Query, String> is temporary for now?

Yes. I have the issue to improve the diagnostics. Which means actually do some diagnostics.

Reply:
To refer the the original argument of "it adds visual noise unrelated to the concept we are explaining" I fail to understand why `.map_err(|e| anyhow!(e))?` helps keep the noise down over `unwrap()` - it adds error handling, a closure and a macro over the ubiquitous `unwrap()` function call.

When the `Result::Err` payload starts implementing the `Error` at some point, I'm happy to switch over to `?` but I'll leave the `unwrap` not to add further noise.

Reply:
We discussed this f2f and the suggestion was about using `anyhow` in general; however, until the error variant here supports the conversion, there's not much we can do to reduce the visual noise, as we're using `anyhow::Result<()>` in the code anyway and `?` just won't work without extra conversion at the moment.

Reply:
Sounds good.